server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java

```diff
@@ -2088,16 +2110,32 @@ && isNotUnitVector(squaredMagnitude)) {
                     }
                 }
             }
+
+            Integer adjustedK = k;
+            int adjustedNumCands = numCands;
+            if (needsRescore(rescoreOversample) && adjustedK != null) {
+                adjustedK = Math.min(OVERSAMPLE_LIMIT, (int) Math.ceil(k * rescoreOversample));
+                adjustedNumCands = Math.max(adjustedK, numCands);
+            }
```

Comment:
Why do we predicate on `adjustedK != null`? 

What if somebody doesn't provide a `k` (like for KnnQueryBuilder)? We should still oversample & rescore on the candidates.

Reply:
I think if `k` isn't provided, it should be oversampled. I think if `k` is provided, that is what we oversample and pick the max between it and num_candidates.

Not oversampling if `k` isn't provided is weird.

Reply:
So we're not oversampling `num_candidates` in case k` `is specified, but we do in case it's not? 🤔 

That sounds a bit weird - I think we should be consistent with how we affect parameters by oversampling so users understand what's happening under the hood related to `num_candidates`.

If `k` is specified, and there's oversampling, we take the top `k * oversample` from `num_candidates` from each shard, rescore them, and get the final best k.

If k` `is not specified, we take `num_candidates` from each shard (as before), and rescore them.

Maybe we need two different parameters here? `rescore_vector` (boolean) and `rescore_k_oversample` (float)?

Reply:
`rescore_oversample: 1.0` indicates a rescore with no oversample. 

I think its equally strange that somebody can do this:

```
"knn": {
 "num_candidates": 50,
 "rescore_vector": {"oversample": 2.0}
}
```

```
"knn": {
 "num_candidates": 50,
 "rescore_vector": {"oversample": 10.0}
}
```

And end up with exactly the same results.

Reply:
This conversation and our desire to eventually have default values for a mapped field for oversampling & rescoring, makes me think the oversampling & rescoring part should be on `num_candidates` and not `k`.

Reply:
What about making the number of rescored vectors per shard explicit, instead of providing a multiplier? Something like:

```json
{
  "k": 10,
  "num_candidates": 100,
  "similarity": 10000000,
  "vector_rescore": {
    "enabled": true,
    "rescore_size": 50
  }
}
```

Where `k` <= `rescore_size` <= `num_candidates`

This way, we're making explicit the number of vectors considered for rescoring. It's straightforward for users to decide whether they want to:

- Rescore all candidates per shard if `rescore_size` is not specified
- Apply rescore to a subset of  the `num_candidates`

I find this easier to reason about, no need to apply a factor to another parameter.

Reply:
@carlosdelest then we cannot set a sane default. What if `k` is greater than rescore size?

How would we configure a default rescoring in the mapping with a static number like this?

Reply:
I get your point - I was thinking on the query and not on the mapping default.

Would it make sense to make the query use a static number like the proposal above, and the mapping default use a factor? It would make defaults depend on `k` (or `num_candidates`, depending on config), and queries use a specific number.

So in mappings we would have something like:

```json
{
  "my_vector": {
    "type": "dense_vector",
    "dims": 64,
    "index_options": {
      "type": "bbq_hnsw",
      "vector_rescore": {
        "k_factor": 3.5
      }
    }
  }
}
```
or alternatively, rescore on a factor of num_candidates using:

```json
{
  "my_vector": {
    "type": "dense_vector",
    "dims": 64,
    "index_options": {
      "type": "bbq_hnsw",
      "vector_rescore": {
        "num_candidates_factor": 3.5
      }
    }
  }
}
```

but the queries could override this by providing exact k, num_candidates and rescore sizse:

```json
{
  "k": 10,
  "num_candidates": 100,
  "similarity": 10000000,
  "vector_rescore": {
    "enabled": true,
    "rescore_size": 50
  }
}
```

Reply:
That is so complicated :(. Is there a particular reason for not doing oversampling & rescoring based off of `num_candidates`?

Reply:
No, there is not 🙂  . I will work on that next.

I was trying to make this more straightforward - but it probably is more complicated combining it with the defaults.

Reply:
@carlosdelest yeah, I think being very clear and calling it "num_candidates_factor" is good. 

Maybe we add various other options in the future, and enforce only one can be configured at a time.